src/tests/mmi.test.ts

```diff
@@ -0,0 +1,244 @@
+
+    const contract = getFlashMintWrappedContract(signer)
+    console.log(contract.address, 'contract')
+    const { inputOutputTokenAmount } = quote
+
+    await wrapETH(inputOutputTokenAmount, signer)
+    await approveErc20(
+      weth.address,
+      contract.address,
+      inputOutputTokenAmount,
+      signer
+    )
+    const allo = await allowanceOf(weth.address, contract.address, signer)
+    console.log(allo.toString())
+
+    const balanceBefore = await balanceOf(signer, mmi.address)
+    console.log(balanceBefore.toString(), 'MMI balance before')
+
+    // Estimate gas
+    const gasEstimate = await contract.estimateGas.issueExactSetFromERC20(
+      mmi.address,
+      weth.address,
+      quote.indexTokenAmount,
+      quote.inputOutputTokenAmount, // _maxAmountInputToken
+      quote.componentSwapData,
+      quote.componentWrapData
+    )
+    console.log(gasEstimate.toString(), 'gasEstimate')
+
+    const tx = await contract.issueExactSetFromERC20(
+      mmi.address,
+      weth.address,
+      quote.indexTokenAmount,
+      quote.inputOutputTokenAmount, // _maxAmountInputToken
+      quote.componentSwapData,
+      quote.componentWrapData,
+      { gasLimit: gasEstimate }
+    )
+    tx.wait()
+
+    const balanceAfter = await balanceOf(signer, mmi.address)
+    console.log(balanceAfter.toString(), 'MMI balance after')
+    expect(balanceAfter.gte(balanceBefore.add(inputOutputTokenAmount))).toBe(
+      true
+    )
+  })
+
+  test('can redeem MMI for USDC', async () => {
+    const indexTokenAmount = wei(1)
+    const outputToken = usdc
+
+    // Mint MMI for having funds to redeem
+    await mintMMI(indexTokenAmount)
+
+    // Get quote
+    const quoteRequest = {
+      isMinting: false,
+      inputToken: mmi,
+      outputToken,
+      indexTokenAmount,
+      slippage: 0.5,
+    }
+    const quoteProvider = new FlashMintQuoteProvider(provider)
+    const quote = await quoteProvider.getQuote(quoteRequest)
+    if (!quote) fail()
+    expect(quote.inputToken).toEqual(mmi)
+    expect(quote.indexTokenAmount).toEqual(indexTokenAmount)
+    expect(quote.inputOutputAmount.gt(0)).toBe(true)
+    expect(quote.slippage).toEqual(quoteRequest.slippage)
+    expect(quote.tx).toBeDefined()
+
+    console.log(indexTokenAmount.toString(), quote.inputOutputAmount.toString())
+
+    const { contract, tx } = quote
+
+    const balanceMMI = await balanceOf(signer, mmi.address)
+    console.log(balanceMMI.toString(), 'balance MMI')
+
+    // Approve spending MMI
+    await approveErc20(mmi.address, contract, indexTokenAmount, signer)
+
+    // Balance before
+    const balanceBefore = await balanceOf(signer, outputToken.address)
+    console.log(balanceBefore.toString(), 'balanceBefore')
+
+    // TODO: Estimate gas fails w/ `ERC20: transfer amount exceeds balance`
+    const gasEstimate = await provider.estimateGas(tx)
+    tx.gasLimit = gasEstimate.mul(120).div(100)
+    console.log(gasEstimate.toString())
```

Comment:
Gas estimate is currently failing. Interestingly enough if commented out and a `gasLimit` is set manually (`6_000_000`) for tx, the tx goes through.